### PR TITLE
Shadow atlas for clustered spot lights

### PR DESCRIPTION
**Pull request by mvaligursky**
This PR builds on https://github.com/playcanvas/engine/pull/2990 to implement shadows for spot clustered lights. 
One of the limitations of the current system is that each shadow casting light needs a shadow map textures, and there is only 16 of them available on most devices. When the material uses few of these, there are not many left for shadows.
This PR address this by allowing all shadow maps for spot lights to be allocated from a single shadow map atlas.

Note: this is still an experimental tech, and is disabled by default.

**Other changes**
- enforcing 0..1 range on shadow bias values on the light, to be in line with the Editor
- new static class floatPacking which handles JS side of packing float into bytes with different size vs quality options. Also moved float2Half here from math. There's a matching .frag chunk to handle the unpacking.
- extracted shadow bias code from multiple locations in forward-render to a single function on the Light, to allow this to be re-used by the clustered lighting. This needs further refactoring in the future. See https://github.com/playcanvas/engine/issues/2780

partial implementation of https://github.com/playcanvas/engine/issues/2691

Screenshot: 64 spot lights with shadow
Screenshot 2021-09-17 at 18 02 55

example (when engine released): https://playcanvas.github.io/#/graphics/clustered-shadows

### CODE REVIEW

**Review by slimbuck — COMMENTED** [review]
More great work! Left some minor comments.

File: src/graphics/program-lib/chunks/clusteredLight.frag [hunk 1]
```diff
@@ -11,90 +18,219 @@ uniform vec3 clusterBoundsDelta;
...
+uniform float shadowAtlasParams;
+
+struct ClusterLightData {
+
+    // v coordinate to look up the light textures
+    float lightV;
+
+    float type;
+    float shape;
+    float falloffMode;
```
[the remark this answers is not in the document]
Comment (mvaligursky) [reply src/graphics/program-lib/chunks/clusteredLight.frag:30]: perhaps. Those are documented on the CPU side when I write to them .. but maybe here as well. In a follow up PR as I've change this file quite a bit for cookies.

File: src/graphics/program-lib/chunks/clusteredLight.frag [hunk 2]
```diff
@@ -11,90 +18,219 @@ uniform vec3 clusterBoundsDelta;
...
+
+    // evaluate omni part of the light
+    getLightDirPoint(light.position);
+    dAtten = getFalloffLinear(light.range);
     if (dAtten > 0.00001) {
 
         // spot light
-        if (lightType > 0.5) {
-            dAtten *= getSpotEffect(lightDir, innerConeAngleCos, outerConeAngleCos);
+        if (light.type > 0.5) {
```
[the remark this answers is not in the document]
Comment (mvaligursky) [reply src/graphics/program-lib/chunks/clusteredLight.frag:158]: I tried bools and they work. I'll try int as well and do that if that works on WebGl1.
In a follow up PR as I've changed this file quite a bit for cookies.

File: src/graphics/program-lib/chunks/clusteredLight.frag [hunk 3]
```diff
@@ -11,90 +18,219 @@ uniform vec3 clusterBoundsDelta;
...
+            decodeClusterLightSpot(light);
+            dAtten *= getSpotEffect(light.direction, light.innerConeAngleCos, light.outerConeAngleCos);
+        }
+
+        // shadow
+        if (light.castShadows > 0.5) {
```
[the remark this answers is not in the document]
Comment (mvaligursky) [reply src/graphics/program-lib/chunks/clusteredLight.frag:164]: yep

File: src/graphics/program-lib/chunks/clusteredLight.frag [hunk 4]
```diff
@@ -11,90 +18,219 @@ uniform vec3 clusterBoundsDelta;
...
+
+            evaluateClusterLight(indices.z);
+
+            if (indices.w <= 0.0)
+                break;
+
+            evaluateClusterLight(indices.w);
+
+            // end of the cell array
+            if (lightCellIndex > clusterPixelsPerCell) {
```
[the remark this answers is not in the document]
Comment (mvaligursky) [reply src/graphics/program-lib/chunks/clusteredLight.frag:231]: webgl1 cannot loop to a dynamic number, only constant.

File: src/graphics/program-lib/chunks/float-unpacking.frag [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+float bytes2float3(vec3 data) {
+    return dot(data, vec3(1.0, 1.0 / 255.0, 1.0 / 65025.0));
+}
+
+float bytes2float4(vec4 data) {
+    return dot(data, vec4(1.0, 1.0 / 255.0, 1.0 / 65025.0, 1.0 / 16581375.0));
+}
+
+float bytes2floatRange2(vec2 data, float min, float max) {
```
[the remark this answers is not in the document]
Comment (mvaligursky) [reply src/graphics/program-lib/chunks/float-unpacking.frag:14]: I've forgotten about the good languages already :(
Thanks for the reminder.
But it's perhaps better to leave it explicit and matching with the CPU packing side.

File: src/math/float-packing.js [hunk 6]
```diff
@@ -0,0 +1,110 @@
+import { math } from './math.js';
+
+let checkRange = 5;
+const oneDiv255 = 1 / 255;
+const floatView = new Float32Array(1);
+const int32View = new Int32Array(floatView.buffer);
+
+// utility static class providing functionality to pack float values to various storage representations
+class floatPacking {
```
[the remark this answers is not in the document]
Comment (mvaligursky) [reply src/math/float-packing.js:9]: yeah, perhaps. In general with use those lower cases name spaces (like math), so I follow this convention .. even thought this is a static class. Maybe should follow class convention here. @willeastcott ?

File: src/scene/lighting/light-texture-atlas.js [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+        const lights = _tempArray;
+        lights.length = 0;
+        for (let i = 0; i < spotLights.length; i++) {
+            const light = spotLights[i];
+            if (light.visibleThisFrame && light.castShadows) {
+                lights.push(light);
+            }
+        }
+
+        if (lights.length > 0) {
```
[the remark this answers is not in the document]
Comment (mvaligursky) [reply src/scene/lighting/light-texture-atlas.js:95]: yep good point, all apart from updateUniforms. I've done it locally for next PR, as this file handles cookies and is different now.

File: src/math/float-packing.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (mvaligursky) [reply src/math/float-packing.js:9]: done in my other PR that builds on this and is ready as well.

File: src/scene/light.js [hunk 8]
```diff
@@ -261,6 +269,43 @@ class Light {
...
+                    tmpVec2.x = -0.00001 * 20;
+                } else {
+                    tmpVec2.x = (this.shadowBias / farClip) * 100;
+                    if (!this.device.webgl2 && this.device.extStandardDerivatives) tmpVec2.x *= -100;
+                }
+                tmpVec2.y = this._isVsm ? this.vsmBias / (farClip / 7.0) : this._normalOffsetBias;
+                break;
+        }
+
+        return tmpVec2;
```
[the remark this answers is not in the document]
Comment (mvaligursky) [reply src/scene/light.js:306]: Same here, doing in the other follow up PR, good idea - as I had to check few times what is in x and what in y :)

**Review by slimbuck — APPROVED** [review]
ship it!

**Pull request merged**